comms/src/parse/Requests.ts

```diff
@@ -0,0 +1,123 @@
+import { Instruction } from "../debug/Instructions";
+
+export interface StackValue {
+    idx: number,
+    type: string;
+    value: number | bigint;
+}
+
+export interface CallbackMapping {
+    callbackid: string;
+    tableIndexes: number[]
+}
+
+export interface InterruptEvent {
+    topic: string;
+    payload: string;
+}
+
+export interface Frame {
+    type: number;
+    fidx: string;
+    sp: number;
+    fp: number;
+    block_key: number;
+    ra: number;
+    idx: number;
+}
+
+export interface Table {
+    max: number;
+    init: number;
+    elements: number[];
+}
+
+export interface Memory {
+    pages: number;
+    max: number;
+    init: number;
+    bytes: Uint8Array;
+}
+
+export interface BRTable {
+    size: string;
+    labels: number[];
+}
+
+export interface State {
+    pc?: number;
+    pc_error?: number;
+    exception_msg?: string;
+    breakpoints?: number[];
+    stack?: StackValue[];
+    callstack?: Frame[];
+    globals?: StackValue[];
+    table?: Table;
+    memory?: Memory;
+    br_table?: BRTable;
+    callbacks?: CallbackMapping[];
+    events?: InterruptEvent[];
+}
+
+
+export interface Request<R> {
+    instruction: Instruction,
+    payload?: string,
+    parser: (input: string) => R,
+    expectedResponse?: (input: string) => boolean
+}
+
+export type RequestBuilder<R> = (payload: string) => Request<R>;
+
+
+const IdentityParser = (line: string) => {
+    return line;
+}
+
+export const RunRequest: Request<string> = {
+    instruction: Instruction.run,
+    expectedResponse: (line: string) => {
+        return line.includes("GO!");
+    },
+    parser: IdentityParser
+}
+
+
+export const PauseRequest: Request<string> = {
+    instruction: Instruction.pause,
+    expectedResponse: (line: string) => {
+        return line.includes("PAUSE!");
+    },
+    parser: IdentityParser
+}
```

Comment:
The `export const PauseRequest: Request<string> =` was just an example. We can indeed make it a boolean.

But now if we remove the `expectedResponse`, it makes no sense to use the `IdentityParser` because it does not throw an error if the string does not match the expected "PAUSE!" ack. This is a problem for the future created for the PauseRequest. The future will always resolve on any given argument to the `PauseRequest.parser`.

Reply:
Ok make it Request<void> and throw an error if includes is false. Basically I would do it like the package already does it but just integrated into the Request:

https://github.com/TOPLLab/WARDuino-libs/blob/250198fd23cc0b27afcf179dc947ea484fe37fe2/comms/src/parse/Parsers.ts#L31-L41

Why do we have to reimplement everything? This is just doing work twice.